### DISCUSSION TITLE
sqlalchemy is very slow comparing to sqlite for fetching a single column

### DISCUSSION THREAD

**Discussion by karlicoss** [sqlalchemy/sqlalchemy#10350] [Usage Questions]
Hi! I've been trying to optimize some of my code that reads from an sqlite database via sqlalchemy. I've reduced it to the following test: read 5 million intergers from an sqlite database.
I attached the self-contained code snippet below.
I'm using latest sqlalchemy 2.0.20, libsqlite version is 3.37.2, python 3.10.4.

Raw `sqlite3` implementation takes 1.2 seconds, whereas when I do it via sqlalchemy core, it takes 3.4 seconds.

After some profiling with `py-spy` I've noticed that most of the time is spent outside sqlite and rather in sqlalchemy itself -- and the time spent in sqlite is about 1 second, which matches the above.

The most offending line performance wise was [this](https://github.com/sqlalchemy/sqlalchemy/blob/8503dc2e948908199cd8ba4e6b1d1ddcf92f4020/lib/sqlalchemy/engine/result.py#L557).
I imagine converting a bare sqlite row to sqlachemy `Row` takes some time, which is fair. However I can't find a way to control that behaviour, e.g. ideally I need a cursor that just returns bare tuples from sqlite without any processing.

I've read the performance [profiling guide too](https://docs.sqlalchemy.org/en/20/faq/performance.html#result-fetching-slowness-core), and it confirms row proxies can be slow, however doesn't really mention any mitigation mechanisms.

From the code I figured out a couple of private methods/attributes I could benefit from:

- it seems that returning raw sqlite3 tuples should be possible if `make_row` becomes `None` as a result of `make_row = self._row_getter` call.
  
  Indeed, if I hack `result._row_getter = None`, it gives a big speedup to the code.

- using `result._raw_row_iterator()` gives a similarly big speedup

  However it doesn't seem to be exposed by anything.

I dug into the sqlalchemy code more and found that `_source_supports_scalars` in some coude paths should cause a `Result` to return raw results. For example, [here](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/engine/result.py#L2190C37-L2190C56).
  
The issue is that the only usages of `_source_supports_scalars` in sqlalchemy that sets it to `True` is in [`sqlalchemy.orm`](https://github.com/sqlalchemy/sqlalchemy/blob/8197f458d3809bcb784777c1a9400fb32f47cc77/lib/sqlalchemy/orm/dynamic.py#L148) or in tests, so I can't figure out how it's meant to be used (if meant to be used with core at all?).

I also seen `engine.raw_connection`, which sort of works, but in my case it's pretty extreme -- I do enjoy other nice things sqlalchemy provides, it's just this one query I want to optimize :) 

Any advice? Thanks!

```python
import gc; gc.disable()

N = 5_000_000
# database with 5000000 ints
db_path = 'ints.sqlite'


def init_db():
    from pathlib import Path
    import shutil
    import sqlite3
    if Path(db_path).exists():
        Path(db_path).unlink()
    with sqlite3.connect(db_path) as conn:
        conn.execute('CREATE TABLE data (value)')
        conn.executemany('INSERT INTO data VALUES (?)', [(i,) for i in range(N)])
    conn.close()


# takes about 1.2s
def test_sqlite():
    import sqlite3
    total = 0
    last = None
    with sqlite3.connect(db_path) as conn:
        for (x,) in conn.execute('SELECT data.value FROM data'):
            total += 1
            last = x
    print(last)
    assert total == N  # just in case


# takes about 3.4s
def test_sqlalchemy():
    import sqlalchemy
    from sqlalchemy import Table, MetaData, Column, text
    engine = sqlalchemy.create_engine(f'sqlite:///{db_path}', echo=True)
    meta = MetaData()
    table_cache = Table('data', meta, Column('value', sqlalchemy.Integer))
    query = table_cache.select()
    total = 0
    last = None
    with engine.connect() as conn:
        result = conn.execute(query)
        # result = conn.execute(query).scalars() # works, but takes even slower, about 3.9s
        # result = result._raw_row_iterator()    # works, makes it run in 1.8 seconds
        # result._row_getter = None              # works, makes it run in 1.9 seconds
        for i, in result:
            total += 1
            last = i
    print(last)
    assert total == N  # just in case


#test_sqlite()
test_sqlalchemy()
```

**Comment by zzzeek** [comment]
hi -

you can probably save about 20% on that very specific test (less on more real-world tests) by giving the result a hint to return all rows in a list:

```py

            for i, in result.all():

```

on my fairly slow machine, adding real timing to your test and including the whole query (even allowing for sqlalchemy's query cache to be not-warmed-up) I get 2.6 for raw cursor and 4.9 for SQLAlchemy, so this is is about 88% slower within what I consider to be a "micro-timing" - fetching a single small column in 5M rows only from a local file.

overall, in Python, things are always quite slow compared to any compiled language like Go or Rust, so considering the large number of iterations this is not a severe performance bump.

this test is also not necessarily a real world case.  if you are fetching 5M rows from PostgreSQL over a network connection and the rows feature a few hundred bytes each, network overhead will likely outweigh a small number of seconds of overhead in row construction.   see below for a test against PG on the network, 1M rows instead of 5M, each row has about 200 bytes.  on my machine I get:

```
Total time for function plain cursor: 1.2595760822296143
Total time for function sqlalchemy: 1.6595776081085205
```

so instead of 100% slower, it's 30% slower.

SQLAlchemy has already been optimizing the Result for years and a large portion of it happens in cython now.      the 1.4/2.0 features added to Result might have worked slightly in the opposite direction but not by very much.

>  it's just this one query I want to optimize :)

use the raw cursor for that one, it's documented at https://docs.sqlalchemy.org/en/20/core/connections.html#working-with-the-dbapi-cursor-directly

```py
import gc

gc.disable()
import time
import contextlib
from sqlalchemy import create_engine, text
from sqlalchemy import Table, MetaData, Column, text, Integer


N = 1000000


@contextlib.contextmanager
def time_fn(name):
    now = time.time()
    yield
    print(f"Total time for function {name}: {time.time() - now}")


def init_db():
    e = create_engine("postgresql://scott:tiger@pg14/test")

    meta = MetaData()
    table_cache = Table("data", meta, Column("value", Integer))

    with e.begin() as conn:
        conn.execute(text("DROP TABLE IF EXISTS data"))
        conn.execute(text("CREATE TABLE data (value varchar(200))"))

        conn.execute(
            table_cache.insert(),
            [
                {
                    "value": f"This is some kind of string value we use to "
                    f"have some actual data to pull over the network: {i}"
                }
                for i in range(N)
            ],
        )


def test_cursor():
    e = create_engine("postgresql://scott:tiger@pg14/test")

    total = 0
    last = None

    with e.connect() as conn:
        cursor = conn.connection.cursor()
        with time_fn("plain cursor"):
            cursor.execute("SELECT data.value FROM data")

            for (x,) in cursor.fetchall():
                total += 1
                last = x
    print(last)
    assert total == N  # just in case


# takes about 3.4s
def test_sqlalchemy():
    e = create_engine("postgresql://scott:tiger@pg14/test")

    total = 0
    last = None

    meta = MetaData()
    table_cache = Table("data", meta, Column("value", Integer))
    query = table_cache.select()

    with e.connect() as conn:
        with time_fn("sqlalchemy"):
            result = conn.execute(query)

            for (x,) in result:
                total += 1
                last = x
    print(last)
    assert total == N  # just in case


init_db()


test_cursor()
test_sqlalchemy()
```

**Reply by CaselIT** [reply]
using `result.scalars().all()` seems to be have fastest performance (trying the op test, I get 6.3s, 5.5s using `result.all()`, 5.1s using `result.scalars().all()`). sqlite is 2.7s on my pc.

@zzzeek when using scalars, would it be possible to avoid creating a ROW at all?

@karlicoss you asked the same question also in the chat. As mentioned there, if these times are a problem for you python is very likely not the right language for your use case, since if you do anything significant your 5m ints the time used by sqlalchemy becomes negligible

**Reply by CaselIT** [reply]
improvement on this are tracked in https://github.com/sqlalchemy/sqlalchemy/issues/10541
